src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,141 @@
+            var resultId = await _tokensCache.GetNextResultIdAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+
+            // Even though we want to ultimately pass edits back to LSP, we still need to compute all semantic tokens,
+            // both for caching purposes and in order to have a baseline comparison when computing the edits.
+            var updatedTokens = await SemanticTokensHelpers.ComputeSemanticTokensAsync(
+                request.TextDocument, resultId, clientName, SolutionProvider,
+                range: null, cancellationToken).ConfigureAwait(false);
+
+            // Getting the cached tokens for the document. If we don't have an applicable cached token set,
+            // we can't calculate edits, so we must return all semantic tokens instead.
+            var cachedTokens = await _tokensCache.GetCachedTokensAsync(
+                request.TextDocument.Uri, request.PreviousResultId, cancellationToken).ConfigureAwait(false);
+            if (cachedTokens == null)
+            {
+                return updatedTokens;
+            }
+
+            // The Data property is always populated on the server side, so it should never be null.
+            Contract.ThrowIfNull(cachedTokens.Data);
+            Contract.ThrowIfNull(updatedTokens.Data);
+
+            var edits = ComputeSemanticTokensEdits(resultId, cachedTokens.Data, updatedTokens.Data);
+            await _tokensCache.UpdateCacheAsync(request.TextDocument.Uri, updatedTokens, cancellationToken).ConfigureAwait(false);
+            return edits;
+        }
+
+        /// <summary>
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdits ComputeSemanticTokensEdits(
+            string resultId,
+            int[] cachedSemanticTokens,
+            int[] updatedSemanticTokens)
+        {
+            using var _1 = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var edits);
+            var index = 0;
+
+            // There are three cases where we might need to create an edit:
+            //     Case 1: Both cached and updated tokens have values at an index, but the tokens don't match
+            //     Case 2: Cached tokens set is longer than updated tokens set - need to make deletion
+            //     Case 3: Updated tokens set is longer than cached tokens set - need to make insertion
+
+            using var _2 = ArrayBuilder<int>.GetInstance(out var currentEdit);
+            while (index < cachedSemanticTokens.Length && index < updatedSemanticTokens.Length)
+            {
+                // Case 1: Both cached and updated tokens have values at index, but the tokens don't match.
+                // We want to make the least number of edits possible, so we keep track of consecutive changes
+                // and report them as a singular edit.
+                if (cachedSemanticTokens[index] != updatedSemanticTokens[index])
```

Comment:
yup, this algorithm still seems untenable.  consdier a simple case of 999 tokens before, and one new token added at the start after.    We start at index 0, and the tokens are different, so we add an edit. tehn we move index 1 and add an edit, and index 2 and so on.  so we end up reporting 999 edits in this case.